src/tests/ftest/cart/test_group_np_srv.c

```diff
@@ -35,6 +35,38 @@
 #include "test_group_rpc.h"
 #include "test_group_np_common.h"
 
+/* Callback to process a SWIM message */
+static void
+swim_crt_event_cb(d_rank_t rank, enum crt_event_source src,
+		  enum crt_event_type type, void *arg)
+{
+
+		/* Example output for SWIM CRT_EVT_DEAD on rank #2:
+		 *	 rank = 2, crt_event_source = 1, crt_event_type = 1
+		 *
+		 *		enum crt_event_type {
+		 *			CRT_EVT_ALIVE,
+		 *			CRT_EVT_DEAD,
+		 *		};
+		 *		enum crt_event_source {
+		 *			CRT_EVS_UNKNOWN,
+		 *			CRT_EVS_SWIM,
+		 *		};
+		 */
+
+		D_DEBUG(DB_TEST, "Cart callback event: "
+			"rank = %d, "
+			"crt_event_source = %d, "
+			"crt_event_type = %d\n",
+			 rank, src, type);
+
+		if (src == 1) {
+			swim_status_by_rank[rank] = type;
```

Comment:
This will only record last even seen which can hide potential bugs if for example server ends up being 'dead, 'alive', 'dead' which it should not do.

Instead it is probably best to keep 2 separate counters: e.g. change swim_status_by_rank to be structure:

struct rank_status {
    uint32_t num_alive;
    uint32_t num_dead;
};

and then based on whether you get alive or dead event increment it e.g.:

if (type == CRT_EVT_ALIVE)
    swim_status_by_rank[rank].alive++;
else if (type == CRT_EVT_DEAD) 
    swim_status_by_rank[rank].dead++;

After that, when you perform verification you verify that number of alives and dead messages matches what you expect.

Reply:
be45e5eb1bd7007dfa9d74e4765bc0ed026f2c5d

Reply:
For valid `rankN=dead` sequences I have this regex: `a*d+` (optional `a`'s to start with one or more ending `d`'s).  For valid `rankN=alive` sequences I have `a*`(no `d`'s).  I don't see how these regex's are achieved via an integer pair of dead/alive counts. (Note: I assume ranks default to an "alive" swim status.) I propose an actual _sequence_ check on swim status.